libraries/AP_HAL_Linux/RCOutput_PCA9685.cpp

```diff
@@ -224,17 +217,17 @@ void RCOutput_PCA9685::push()
         *d++ = length >> 8;
     }
 
-    if (!_i2c_sem->take_nonblocking()) {
+    if (!_dev->get_semaphore()->take_nonblocking()) {
         return;
     }
 
-    hal.i2c->writeRegisters(_addr,
-                            PCA9685_RA_LED0_ON_L + 4 * (_channel_offset + min_ch),
-                            (max_ch - min_ch) * 4,
-                            &data[min_ch * 4]);
-
-    _i2c_sem->give();
+    size_t len = (max_ch - min_ch) * 4 + 1;
+    uint8_t transfer_data[len];
+    transfer_data[0] = PCA9685_RA_LED0_ON_L + 4 * (_channel_offset + min_ch);
+    memcpy(&transfer_data[1], &data[min_ch * 4], (len - 1) * sizeof(uint8_t));
```

Comment:
I'd prefer not to copy this data. What we can do is to change the `data` declaration above to be:

```
struct pwm_values {
    uint8_t reg;
    uint8_t val[PWM_CHAN_COUNT * 4];
} pwm_values;
```

If you look at the for loop above, we always write from the beginning of the array. This way if we are updating from ch0 to ch4 we have:

```
[PCA9685_RA_LED0_ON_L + 4 * (_channel_offset)][0][0][ch0_low][ch0_high][0][0][ch1_low][ch1_high][0][0][ch2_low][ch2_high][0][0][ch3_low][ch3_high]
```

From ch1 to ch4:
```
[PCA9685_RA_LED0_ON_L + 4 * (_channel_offset + 1)][ch1_high][0][0][ch2_low][ch2_high][0][0][ch3_low][ch3_high]
```

`len` the way you did continues to be correct.  Think again I think it would be much clearer if the values were declared as uint16_t